### DISCUSSION TITLE
modelling genetically correlated traits

### DISCUSSION THREAD

**Discussion by mahlet-teka** [gaynorr/AlphaSimR#97] [Q&A] [answered]
I have setup the SP and used the SP$addTraitA to generate the traits;

   SP$addTraitA(nQtlPerChr=nQTL, mean = c(85, 200, 40, 850, 85, 200, 40, 850), var = Vg, corA=rho)    
   
 # add error variance as function of heritability
    SP$setVarE(h2 = founderh2)
    
    # add genotype platform (i.e add/sample markers)
    SP$addSnpChip(nM)
   founders <- newPop(founderPop)
 
rho is an 8X8 correlation matrix where the correlation between some of the traits (such as trait 1 and 5) is really high (0.9). But the when I try to see if indeed the genetic correlation between trait 1 and trait 5 is really 0.9, I used;
cor(founders@gv[,1], founders@gv[,5]), and the correlation is way lower than 0.9.

why is that?

I appreciate your help!
Mahlet.

**Comment by gregorgorjanc** [comment, marked as answer]
@mahlet-teka this can indeed happen, but it will be highly dependent on sample size - number of individuals, number of chromosomes and QTL. If you simulate few individuals, chromosomes, and QTL you can get quite a bit of deviation from the targeted correlation. Maybe you expand your code with just two traits and show your numbers of chromosomes, QTL, and individuals. I think this is an important example that can benefit others in the community too.

**Reply by mahlet-teka** [reply]
@gregorgorjanc I have simulated two traits with the parameters below and the genetic correlation seems to be what I setup. But when I expanded to my situation, which is 8 traits, the correlation is not what it should be. Here is the simulation and the correlation results. 

###parameters
nChrom <- 10  
nLoci <- 100  
nM <- 25      
nQTL <- 25  

Vg <- c(5.97, 105.9, 41.75, 11421, 6.4, 109.1, 45.00, 10001)    
founderh2 <- c(0.45, 0.25, 0.81, 0.50, 0.37, 0.20, 0.85, 0.50)  
Ve <- (1 - founderh2) / founderh2 * Vg           
nFounder = 1200                                 

rho =    [,1]      [,2]         [,3]            [,4]         [,5]    [,6]           [,7]         [,8]
  [1,]  1.0000 2.0e-01  0.710000 -0.1700  **0.9000** 4.0e-02  0.504100  0.0289
  [2,]  0.2000 1.0e+00  0.006000  0.7000  0.0400 9.0e-01  0.000036  0.4900
  [3,]  0.7100 6.0e-03  1.000000 -0.6900  0.5041 3.6e-05  0.900000  0.4761
  [4,] -0.1700 7.0e-01 -0.690000  1.0000  0.0289 4.9e-01  0.476100  0.9000
  [5,]  0.9000 4.0e-02  0.504100  0.0289  1.0000 2.0e-01  0.710000 -0.1700
  [6,]  0.0400 9.0e-01  0.000036  0.4900  0.2000 1.0e+00  0.006000  0.7000
  [7,]  0.5041 3.6e-05  0.900000  0.4761  0.7100 6.0e-03  1.000000 -0.6900
  [8,]  0.0289 4.9e-01  0.476100  0.9000 -0.1700 7.0e-01 -0.690000  1.0000

founderPop <- quickHaplo(nFounder, nChrom, nLoci, genLen = 1, ploidy = 2L, inbred = FALSE)
    
# Setting Simulation Parameters
SP <- SimParam$new(founderPop)
   
# add multi-trait (i.e. add/sample qtl)
SP$addTraitA(nQtlPerChr=nQTL, mean = c(85, 85), var = Vg, corA=rho) 
    
# add error variance as function of heritability
 SP$setVarE(h2 = founderh2)
    
# add genotype platform (i.e add/sample markers)
 SP$addSnpChip(nM)
    
founders <- newPop(founderPop) 

founders
An object of class "Pop" 
Ploidy: 2 
Individuals: 1200 
Chromosomes: 10 
Loci: 100000 
Traits: 8 

#########
 - as you can see from the correlation matrix rho, the correlation between trait 1 and 5; 2 and 6; 3 and 7; and 4 and 8; is 0.9. But when I try to calculate the correlation as;

> cor(founders@gv[,1], founders@gv[,5])
   0.7799584
> cor(founders@gv[,2], founders@gv[,6])
  0.7668488
> cor(founders@gv[,3], founders@gv[,7])
  0.2788082
> cor(founders@gv[,4], founders@gv[,8])
  0.1950979 

- which is lower than 0.9 and way lower for the last two traits.

**Reply by gregorgorjanc** [reply]
@mahlet-teka is the corresponding covariance matrix positive-definite? Simple test is to see if you can invert it - use solve().

**Reply by mahlet-teka** [reply]
Yes, I can invert it. but when I check is.positive.definate(rho), its seems its not.

**Reply by gregorgorjanc** [reply]
Ok, this is the problem. You need a valid covariance matrix. AlphaSimR internally “fixes” the matrix, but as you see this can impact the output of simulation.

@gaynorr we need to throw a warning in such cases so users are aware - will open an issue to track this.

**Reply by mahlet-teka** [reply]
Alright. I will fix it and try with the correct matrix.

**Reply by gregorgorjanc** [reply]
@mahlet-teka can you tag/mark one of the above posts as an answer so other people are aware of what to look for?
